fix(cli): 修复沙箱僵尸进程累积导致资源损耗

### 问题摘要 / Issue Summary

修复沙箱僵尸进程累积导致资源损耗

### 版本信息 / Version Information

N/A

### 运行环境 / Runtime Environment

N/A

### 发生了什么？/ What happened?

跟踪并修复 agent-infra 长期运行沙箱未回收孤儿子进程的问题。当前任务沙箱使用 tail -f /dev/null 作为 PID 1，且创建容器时未启用 init；反复运行 build、测试和 detached 子进程后，退出的 node、sh、tsc 等进程被 PID 1 接管但未 wait，形成大量僵尸进程并占用 PID 等系统资源。需要减少这类资源损耗，并验证沙箱执行构建、测试和进程终止场景后不会持续累积僵尸进程。

### 期望的行为 / Expected Behavior

N/A

### 重现步骤 / Steps to Reproduce

N/A

### 相关日志 / Relevant Logs

N/A

### 额外信息 / Additional Context

N/A

### 任务输入

<!-- 由 create-task 忠实记录当前请求及必要前序讨论中已有的信息；缺失项保持为空，不推导补齐。 -->

### 来源

- 用户当前请求：创建任务跟踪当前沙箱僵尸进程情况，以减少资源损耗。
- 当前会话中的宿主机与容器只读取证结果。

### 已确认事实与证据

- 容器 agent-infra-dev-agent-infra-bugfix-preserve-artifact-timeline-on-sync 中观测到 153 个僵尸进程：122 个 node、24 个 sh、4 个 tsc，以及 git、gh、cc-token-status 各 1 个。
- 所有僵尸进程的当前父进程均为容器 PID 1。
- 容器 PID 1 的命令为 tail -f /dev/null，Docker HostConfig.Init 为 null。
- 僵尸产生集中在任务实现和代码审查运行 build、定向测试及 npm test 的时间窗口。
- 退出状态包含正常退出、状态码 1、SIGTERM 和 SIGKILL。
- 现有僵尸基本不消耗 CPU 或实际内存，但占用 PID；当前高 CPU 主要来自存活的 OrbStack、Codex 和 Node 工作负载。

### 约束

- 保持现有沙箱工作树和挂载数据持久化语义。
- 不得依赖删除测试或取消进程树终止覆盖来规避问题。
- install.sh 如被修改必须保持 POSIX sh 兼容。
- 代码、CLI 文本和标识符使用英文；项目文档变更需提供英文主版本与中文翻译。

### 已确认决策

- 用户已决定创建独立任务跟踪该资源损耗问题。

### 候选与否决方案

- Agent 建议候选方案：创建容器时启用 Docker --init 或使用 tini 作为 PID 1；具体方案需在后续分析和设计阶段确认。
- 仅重启容器可以清理当前僵尸，但不能防止后续再次累积，因此不作为根治方案。

### 验收标准

- 在沙箱内运行项目 build、测试以及包含 detached 子进程终止的场景后，已退出子进程能够被回收，容器僵尸进程数量不会随重复执行持续增长。
- 修复后现有沙箱创建、进入、任务绑定、挂载和工作树持久化行为保持可用。
- 提供可重复的验证，覆盖正常退出、非零退出、SIGTERM 和 SIGKILL 子进程的回收行为。

### 未决事项

- 后续分析需确认由 Docker --init、镜像内 tini，还是其他等价 init/reaper 机制承担 PID 1 回收职责。

<!-- sync-issue:TASK-20260819-231453:code-r4 -->
## 实现报告（Round 4）

> **codex** · TASK-20260819-231453

# 实现报告

- **实现轮次**: Round 4
- **产物文件**: `code-r4.md`

## 实现输入

- **模式**：fix
- **方案输入**：`plan.md`
- **审查输入**：`review-code-r4.md`
- **裁决输入**：N/A
- **账本 ID**：N/A
- **裁决证据**：N/A
- **需求摘要**：修复显式 `ai sandbox start --recreate` 在健康运行容器上仍然 no-op 的 blocker，使存量容器无论 readiness 是否健康都进入既有受保护的 container-only replacement 边界；普通 `start` 保持健康 no-op。

## 状态核对

```text
$ agent-infra-internal task-snapshot TASK-20260819-231453 --format text
$ git status -s
(empty)
$ ls -la .agents/workspace/active/TASK-20260819-231453/
f 6793 analysis.md
f 8202 code-r2.md
f 6473 code-r3.md
f 6239 code.md
f 8220 plan.md
f 15553 review-analysis-r2.md
f 15443 review-analysis.md
f 19601 review-code-r2.md
f 18677 review-code-r3.md
f 12533 review-code-r4.md
f 19199 review-code.md
f 18697 review-plan.md
f 11973 task.md
f 7152 validation-run.md
$ tail .agents/workspace/active/TASK-20260819-231453/task.md
- 2026-08-20 23:38:36+08:00 — **Review Code (Round 4)** by codex — Verdict: Changes Requested, blockers: 1, major: 0, minor: 0, Manual-validation: 1 → review-code-r4.md
```

## 变更文件

### 新建文件

- `.agents/workspace/active/TASK-20260819-231453/code-r4.md` - 本轮修复报告。

### 修改文件

- `lib/sandbox/commands/start.ts` - 将显式 `--recreate` 作为强制重建意图传给 readiness/replacement 边界，并更新 CLI 帮助语义。
- `lib/sandbox/recovery.ts` - 新增局部 `forceRecreate` 输入，在 broker readiness 后跳过健康/恢复早退，复用既有工作树快照、stop/rm/create、替换后 readiness 和漂移保护。
- `tests/integration/cli/sandbox-tmpfs.test.ts` - 新增健康运行容器显式重建回归，验证 stop、rm、create callback 与 `path: recreated`。

## 关键代码说明

### 显式重建进入既有安全替换边界

**文件**: `lib/sandbox/recovery.ts:1071`

**实现逻辑**:

`start` 仅在用户显式传入 `--recreate` 时同时设置 `allowRecreate` 和 `forceRecreate`。readiness 仍先完成 control broker 协调，随后强制意图通过既有 catch 路径进入 replacement；因此不会复制 stop/rm/create，也不会绕过工作树 identity 快照、替换后 readiness 或漂移检查。未传 flag 的 `start`、`exec --recreate` 和 capture recovery 语义不变。

**关键代码**:

```typescript
if (params.forceRecreate) {
  throw new Error('Explicit container recreation requested.');
}
```

### 命令入口只对 start 启用强制语义

**文件**: `lib/sandbox/commands/start.ts:79`

**实现逻辑**:

`allowRecreate` 继续表示“恢复失败时允许替换”，`forceRecreate` 明确表示“本次 start 就要求替换”。两个信号在 start 的显式 flag 下同时为真，保持其他调用方只授权失败恢复、不无条件重建。

## 测试结果

### RED-GREEN 回归

- 测试文件：`tests/integration/cli/sandbox-tmpfs.test.ts`
- 新增用例数：1
- RED：旧实现 25 项中 24 通过、1 失败；`recreated` 实际为 false。
- GREEN：修复后 25/25 通过。

### 项目验证

- `npm run typecheck`：通过。
- `npm run test:smoke`：1248 项，1243 通过，0 失败，5 跳过。
- `npm run test:core`：2110 项，2102 通过，0 失败，8 跳过。
- `git diff --check`：通过。

**测试输出**:

```text
RED: tests 25 | pass 24 | fail 1 | explicit recreation replaces a healthy running container
GREEN: tests 25 | pass 25 | fail 0
SMOKE: tests 1248 | pass 1243 | fail 0 | skipped 5
CORE: tests 2110 | pass 2102 | fail 0 | skipped 8
```

## 证据原文

- 断言：新增用例先稳定复现旧行为。

```text
$ node scripts/run-tests.js tests/integration/cli/sandbox-tmpfs.test.ts
ℹ tests 25
ℹ pass 24
ℹ fail 1
✖ explicit recreation replaces a healthy running container
AssertionError: false !== true
```

- 断言：最小修复后同一用例转绿，且类型检查与 smoke 通过。

```text
$ node scripts/run-tests.js tests/integration/cli/sandbox-tmpfs.test.ts && npm run typecheck && npm run test:smoke
ℹ tests 25
ℹ pass 25
ℹ fail 0
> tsc -p tsconfig.test.json --noEmit && tsc -p tsconfig.jschecks.json --noEmit
ℹ tests 1248
ℹ pass 1243
ℹ fail 0
ℹ skipped 5
```

- 断言：最终 core 门禁与 diff 格式检查通过，业务修改限定 3 个文件。

```text
$ npm run test:core && git diff --check
ℹ tests 2110
ℹ pass 2102
ℹ fail 0
ℹ skipped 8
$ git status --short
 M lib/sandbox/commands/start.ts
 M lib/sandbox/recovery.ts
 M tests/integration/cli/sandbox-tmpfs.test.ts
```

- 断言：显式重建复用既有替换命令顺序并返回 recreated。

```text
$ nl -ba tests/integration/cli/sandbox-tmpfs.test.ts | sed -n '800,839p'
800 test("explicit recreation replaces a healthy running container", async () => {
813 allowRecreate: true,
814 forceRecreate: true,
831 assert.deepEqual(replacementCommands, [
832   ["stop", "demo-dev-feature..demo"],
833   ["rm", "demo-dev-feature..demo"]
835 assert.equal(result.path, "recreated");
```

## 与方案的差异

无功能方向偏离。方案已明确存量容器通过显式 `start --recreate` 收敛；本轮只补足此前遗漏的健康状态转换，并继续复用批准方案中的统一 create 和既有 replacement 安全边界。

账本握手的 finding 来自补充 `review-code-r4.md`，而最新 code 仍为 Round 3，因此核心 `finding-respond` 要求握手轮次从 4 增至 5；实现产物仍按 `task-artifact` 的权威输出使用 `code-r4.md`。首次按实现轮次 4 响应被核心无写入拒绝，随后按账本状态机以 round 5 成功记录 accepted。

## 对审查发现的逐条核实

| 发现 | 处置状态 | 相称证据 |
|------|----------|----------|
| CD-3 / blocker：健康运行容器下显式 `start --recreate` 不替换 | accepted | `lib/sandbox/commands/start.ts:79-90` 传入强制意图；`lib/sandbox/recovery.ts:1071-1073` 进入既有 replacement；`tests/integration/cli/sandbox-tmpfs.test.ts:800-839` RED→GREEN 验证 stop/rm/recreate。 |

## 人工校验项处理

### 修复后真实 Docker 迁移与跨宿主矩阵

该项不在本轮 AI 代码修复范围内。自动化已经验证状态转换与安全替换边界；仍需在真实 Docker 宿主确认健康容器 ID 变化、`HostConfig.Init=true`、工作树/任务绑定/bind mount/`/share`/控制入口保持，以及 Linux native rootful/rootless、Debian 和其他可用引擎矩阵。

## 供审查关注的内容

- `forceRecreate` 是否仅由 `start --recreate` 设置，普通 start、exec/capture 的失败后授权语义是否保持。
- 强制意图是否确实经过 control broker 协调与既有工作树快照/漂移保护，而非复制另一套删除逻辑。
- RED 用例是否充分证明健康路径不再提前返回。

## 已知问题

- 真实 Docker 修复后迁移与跨宿主矩阵仍需后续人工校验；不以 fixture 替代该证据。

## 下一步

重新运行 `review-code`，复核 CD-3 修复、完整状态流、风险镜头和测试证据。未执行 `git add` 或 `git commit`。


---
*由 codex 自动生成 · 内部追踪：TASK-20260819-231453*

--- a/.github/workflows/sandbox-linux-e2e.yml
+++ b/.github/workflows/sandbox-linux-e2e.yml
@@ -142,6 +142,45 @@ jobs:
           container="$(docker ps --filter "label=linux-e2e.sandbox.branch=test-branch" --format '{{.Names}}' | head -n 1)"
           test -n "$container"
           docker exec "$container" sh -lc 'echo hello && node --version'
+          test "$(docker inspect --format '{{.HostConfig.Init}}' "$container")" = true
+          zombie_count() {
+            docker exec "$container" sh -lc 'count=0; for stat in /proc/[0-9]*/stat; do if [ -r "$stat" ]; then state="$(awk '\''{print $3}'\'' "$stat" 2>/dev/null || true)"; [ "$state" = Z ] && count=$((count + 1)); fi; done; printf "%s\n" "$count"'
+          }
+          run_orphan_probe() {
+            docker exec -i "$container" node - "$1" <<'NODE'
+          const { spawn } = require("node:child_process");
+          const mode = process.argv[2];
+          const child = spawn(process.execPath, ["-e", `
+          const mode = process.argv[1];
+          if (mode === "term" || mode === "kill") {
+            setTimeout(() => process.kill(process.pid, mode === "term" ? "SIGTERM" : "SIGKILL"), 100);
+          } else {
+            process.exit(mode === "nonzero" ? 23 : 0);
+          }
+          setTimeout(() => {}, 1000);
+          `, mode], { detached: true, stdio: "ignore" });
+          child.unref();
+          process.stdout.write(String(child.pid) + "\n");
+          NODE
+          }
+          probe_pending() {
+            docker exec "$container" sh -lc 'pending=0; for pid in "$@"; do stat="/proc/$pid/stat"; if [ -r "$stat" ]; then state="$(awk '\''{print $3}'\'' "$stat" 2>/dev/null || true)"; [ "$state" = Z ] || pending=1; fi; done; printf "%s\n" "$pending"' sh "$@"
+          }
+          baseline="$(zombie_count)"
+          for mode in normal nonzero term kill; do
+            probe_pids=""
+            for _ in $(seq 1 10); do probe_pids="$probe_pids $(run_orphan_probe "$mode")"; done
+            pending=1
+            for _ in $(seq 1 20); do
+              pending="$(probe_pending $probe_pids)"
+              final="$(zombie_count)"
+              [ "$pending" -eq 0 ] && [ "$final" -le "$baseline" ] && break
+              sleep 0.1
+            done
+            test "$pending" -eq 0
+            test "$final" -le "$baseline"
+            echo "zombie probe mode=$mode baseline=$baseline final=$final"
+          done
 
       - name: Cleanup sandbox artifacts
         if: always()
@@ -257,6 +296,45 @@ jobs:
           container="$(docker ps --filter "label=linux-e2e.sandbox.branch=test-branch" --format '{{.Names}}' | head -n 1)"
           test -n "$container"
           docker exec "$container" sh -lc 'echo hello && node --version'
+          test "$(docker inspect --format '{{.HostConfig.Init}}' "$container")" = true
+          zombie_count() {
+            docker exec "$container" sh -lc 'count=0; for stat in /proc/[0-9]*/stat; do if [ -r "$stat" ]; then state="$(awk '\''{print $3}'\'' "$stat" 2>/dev/null || true)"; [ "$state" = Z ] && count=$((count + 1)); fi; done; printf "%s\n" "$count"'
+          }
+          run_orphan_probe() {
+            docker exec -i "$container" node - "$1" <<'NODE'
+          const { spawn } = require("node:child_process");
+          const mode = process.argv[2];
+          const child = spawn(process.execPath, ["-e", `
+          const mode = process.argv[1];
+          if (mode === "term" || mode === "kill") {
+            setTimeout(() => process.kill(process.pid, mode === "term" ? "SIGTERM" : "SIGKILL"), 100);
+          } else {
+            process.exit(mode === "nonzero" ? 23 : 0);
+          }
+          setTimeout(() => {}, 1000);
+          `, mode], { detached: true, stdio: "ignore" });
+          child.unref();
+          process.stdout.write(String(child.pid) + "\n");
+          NODE
+          }
+          probe_pending() {
+            docker exec "$container" sh -lc 'pending=0; for pid in "$@"; do stat="/proc/$pid/stat"; if [ -r "$stat" ]; then state="$(awk '\''{print $3}'\'' "$stat" 2>/dev/null || true)"; [ "$state" = Z ] || pending=1; fi; done; printf "%s\n" "$pending"' sh "$@"
+          }
+          baseline="$(zombie_count)"
+          for mode in normal nonzero term kill; do
+            probe_pids=""
+            for _ in $(seq 1 10); do probe_pids="$probe_pids $(run_orphan_probe "$mode")"; done
+            pending=1
+            for _ in $(seq 1 20); do
+              pending="$(probe_pending $probe_pids)"
+              final="$(zombie_count)"
+              [ "$pending" -eq 0 ] && [ "$final" -le "$baseline" ] && break
+              sleep 0.1
+            done
+            test "$pending" -eq 0
+            test "$final" -le "$baseline"
+            echo "zombie probe mode=$mode baseline=$baseline final=$final"
+          done
 
       - name: Cleanup sandbox artifacts
         if: always()

--- a/docs/en/sandbox.md
+++ b/docs/en/sandbox.md
@@ -14,6 +14,16 @@ If this check fails, `ai sandbox create` and `ai sandbox rebuild` stop before `d
 
 The built-in image also verifies that `cc-token-status`, `sandbox-dotfiles-link`, and `sandbox-tmux-entry` are non-empty and executable during the Docker build. A build cannot succeed with the empty scripts produced by a legacy builder. Custom Dockerfiles still require BuildKit, but they are not required to contain these built-in scripts.
 
+## PID 1 and orphan process reaping
+
+Every newly created sandbox enables Docker's managed init with `docker run --init`. The init process becomes PID 1, forwards signals, and reaps orphaned child processes so repeated builds, tests, and detached work do not accumulate zombies. This setting applies when a container is created or explicitly recreated; existing containers retain their previous PID 1 until you run:
+
+```bash
+ai sandbox start --recreate <task-ref-or-branch>
+```
+
+Recreation preserves the worktree, task binding, mounts, and `/share` data while replacing the container-local process state.
+
 ## Sandbox aliases and GitHub CLI
 
 `ai sandbox create` now bootstraps the host-side aliases file at `~/.agent-infra/aliases/sandbox.sh` on first run. The generated file includes ready-to-edit yolo shortcuts for Claude, Codex, Antigravity CLI, and OpenCode, and every sandbox syncs that file into `/home/devuser/.bash_aliases`.

--- a/docs/zh-CN/sandbox.md
+++ b/docs/zh-CN/sandbox.md
@@ -14,6 +14,16 @@ docker buildx inspect --bootstrap
 
 内置镜像还会在 Docker 构建期间验证 `cc-token-status`、`sandbox-dotfiles-link` 与 `sandbox-tmux-entry` 均非空且可执行，因此 legacy builder 生成空脚本时构建不会成功。自定义 Dockerfile 仍需满足 BuildKit 前置条件，但不要求包含这三个内置脚本。
 
+## PID 1 与孤儿进程回收
+
+每个新建沙箱都会启用 Docker 托管的 init（`docker run --init`）。该 init 进程作为 PID 1 负责转发信号并回收孤儿子进程，避免反复运行构建、测试和 detached 工作时持续累积僵尸进程。此设置只在创建或显式重建容器时生效；存量容器会继续使用原来的 PID 1，需执行以下命令重建：
+
+```bash
+ai sandbox start --recreate <task-ref-or-branch>
+```
+
+重建会保留工作树、任务绑定、挂载和 `/share` 数据，但会替换容器内的进程状态。
+
 ## 沙箱 aliases 与 GitHub CLI
 
 `ai sandbox create` 在首次运行时会自动生成宿主机侧的 `~/.agent-infra/aliases/sandbox.sh`。该文件内置了 Claude、Codex、Antigravity CLI 和 OpenCode 的 yolo 快捷命令模板，你可以直接修改；每次创建沙箱时，这个文件都会同步到容器内的 `/home/devuser/.bash_aliases`。

--- a/lib/sandbox/commands/create.ts
+++ b/lib/sandbox/commands/create.ts
@@ -1425,6 +1425,7 @@ export async function create(args: string[]): Promise<void> {
             runEngineTaskCommand(engine, 'docker', [
               'run',
               '-d',
+              '--init',
               '--name',
               container,
               '--hostname',

--- a/lib/sandbox/commands/start.ts
+++ b/lib/sandbox/commands/start.ts
@@ -17,9 +17,9 @@ const USAGE = `Usage: ai sandbox start [--recreate] <branch | TASK-id | N>
 
 Start an existing sandbox container that has stopped (for example after the
 Docker daemon was restarted or replaced). The container must already exist:
-if none is found, run 'ai sandbox create <branch>' first. A container that is
-already running is checked without modifying healthy runtime state. --recreate
-allows container-only replacement after in-place recovery fails.`;
+if none is found, run 'ai sandbox create <branch>' first. Without --recreate, a
+container that is already running is checked without modifying healthy runtime
+state. --recreate replaces the existing container even when it is healthy.`;
 
 export function parseStartArgs(args: string[]): { target: string; recreate: boolean; help: boolean } {
   let recreate = false;
@@ -83,6 +83,7 @@ export async function start(args: string[]): Promise<void> {
     workspace: target.workspace,
     row: found,
     allowRecreate: parsed.recreate,
+    forceRecreate: parsed.recreate,
     recreate: async () => {
       const { create } = await import('./create.ts');
       await create([target.requestedRef, '--no-refresh']);

--- a/lib/sandbox/recovery.ts
+++ b/lib/sandbox/recovery.ts
@@ -170,6 +170,7 @@ type EnsureSandboxReadyParams = {
   workspace?: SandboxWorkspaceIdentity;
   row: SandboxRow;
   allowRecreate?: boolean;
+  forceRecreate?: boolean;
   recreate?: (branch: string) => Promise<void>;
   writeWarning?: (message: string) => void;
   deps?: RecoveryCommandDeps;
@@ -1067,6 +1068,9 @@ export async function ensureSandboxReady(params: EnsureSandboxReadyParams): Prom
         workspace: params.workspace ?? { mode: 'branch-only' }
       });
     }
+    if (params.forceRecreate) {
+      throw new Error('Explicit container recreation requested.');
+    }
     if (!params.row.running) {
       startFn(params.engine, params.row.name);
       const initial = assess({
@@ -1224,7 +1228,9 @@ export async function ensureSandboxReady(params: EnsureSandboxReadyParams): Prom
     );
   }
 
-  const warning = `Sandbox recovery failed in place. Replacing only the container. ${dataBoundary}`;
+  const warning = params.forceRecreate
+    ? `Explicit sandbox recreation requested. Replacing only the container. ${dataBoundary}`
+    : `Sandbox recovery failed in place. Replacing only the container. ${dataBoundary}`;
   warnings.push(warning);
   (params.writeWarning ?? ((message) => process.stderr.write(`${message}\n`)))(warning);
   const runVerboseFn = deps?.runVerbose ?? runVerboseEngine;

--- a/tests/integration/cli/sandbox-core.test.ts
+++ b/tests/integration/cli/sandbox-core.test.ts
@@ -545,6 +545,7 @@ test("sandbox create keeps a clean runtime-only workspace and does not mount the
     const calls = fixture.readDockerCalls();
     const runCall = calls.find((call) => call[0] === "run");
     assert.ok(runCall, "expected sandbox create to invoke docker run");
+    assert.equal(runCall.filter((arg) => arg === "--init").length, 1);
     assert.equal(
       runCall.some((arg) => arg.includes("/home/devuser/.ssh")),
       false,
@@ -603,6 +604,7 @@ test("task-bound sandbox create keeps Git clean and exposes only the scoped writ
     assert.equal(result.status, 0, result.stderr);
     const runCall = fixture.readDockerCalls().find((call) => call[0] === "run");
     assert.ok(runCall, "expected task-bound sandbox create to invoke docker run");
+    assert.equal(runCall.filter((arg) => arg === "--init").length, 1);
     assert.ok(runCall.some((arg) => isReadOnlyMountFor(arg, "/workspace/.agents/workspace")));
     assert.ok(runCall.some((arg) => isWritableMountFor(
       arg,

--- a/tests/integration/cli/sandbox-tmpfs.test.ts
+++ b/tests/integration/cli/sandbox-tmpfs.test.ts
@@ -797,6 +797,47 @@ test("container replacement snapshots the worktree before Docker writes and reje
   }
 });
 
+test("explicit recreation replaces a healthy running container", async () => {
+  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "agent-infra-recovery-explicit-recreate-"));
+  const config = recoveryFixtureConfig(tmpDir);
+  const replacementCommands: string[][] = [];
+  let recreated = false;
+  const row = { name: "demo-dev-feature..demo", status: "Up", branch: "feature/demo", running: true, index: 1 };
+
+  try {
+    const result = await ensureSandboxReady({
+      config,
+      engine: "native",
+      branch: "feature/demo",
+      row,
+      allowRecreate: true,
+      forceRecreate: true,
+      recreate: async () => { recreated = true; },
+      writeWarning: () => {},
+      deps: {
+        ensureControlBroker: async () => {},
+        run: () => JSON.stringify([{
+          Id: "fixture-container-id",
+          Config: { Labels: BRANCH_ONLY_LABELS },
+          Mounts: recoveryFixtureMounts(config)
+        }]),
+        runOk: () => true,
+        runVerbose: (_engine, _cmd, args) => { replacementCommands.push(args); },
+        fetchRows: () => ({ running: [row], nonRunning: [] })
+      }
+    });
+
+    assert.equal(recreated, true);
+    assert.deepEqual(replacementCommands, [
+      ["stop", "demo-dev-feature..demo"],
+      ["rm", "demo-dev-feature..demo"]
+    ]);
+    assert.equal(result.path, "recreated");
+  } finally {
+    fs.rmSync(tmpDir, { recursive: true, force: true });
+  }
+});
+
 test("container replacement preserves the original failure when the worktree is unchanged", async () => {
   const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "agent-infra-recovery-original-failure-"));
   const config = recoveryFixtureConfig(tmpDir);

